.github/workflows/main.yml

```diff
@@ -1,50 +1,120 @@
+          tree=$repo/tree/$GITHUB_REF_NAME
+          ./github-action-pylint-report.py xcp pylint-report.md $tree
+          message_count=`tail -1 pylint-report.md|cut -d'|' -f2`
+          if [ -e $GITHUB_STEP_SUMMARY ]; then  #  The "act" local runner does not have this:
+              echo -n "### [$tool]($tree/$tool) on **xcp/\\*\\*/*.py**" >>$GITHUB_STEP_SUMMARY
+              echo ": #Total Pylint messages: $message_count :rocket: " >>$GITHUB_STEP_SUMMARY
+              cat pylint-report.md                                      >>$GITHUB_STEP_SUMMARY
+          fi
+
+      - name: Run pytest with --cov and generate HTML coverage reports
+        id: coverage
+        if: ${{ matrix.python-version == '3.11' }}
         run: |
-          pytest --cov -rP
+          set -x;PYTHONPATH=$PWD pytest --cov
           coverage xml
-          coverage html
-          coverage html -d htmlcov-tests --include="tests/*"
-          diff-cover --html-report coverage-diff.html coverage.xml
+          if [ -n "$COVERALLS_REPO_TOKEN" ]; then coveralls --submit=coverage.xml; fi
+          coverage html -d htmlcov-${{ matrix.python-version }}
+          coverage html -d htmlcov-tests-${{ matrix.python-version }} --include="tests/*"
+          diff-cover --compare-branch=origin/master --html-report coverage-diff-${{ matrix.python-version }}.html coverage.xml
+        env:
+          COVERALLS_REPO_TOKEN: ${{ secrets.COVERALLS_REPO_TOKEN }}
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: "Run tox for ${{ matrix.python-version }} (and also mypy if 3.11)"
+        if: ${{ matrix.python-version != '2.7' }}  # On Ubuntu 20.04 Python2.7, tox is older
+        run: |
+          set -ex;pip install --quiet --user --upgrade poetry tox tox-gh-actions
+          ~/.local/bin/tox  # Runs pytest & mypy in defined environments, simple to run locally
 
       - name: Pylint
         run: |
           pylint --version
           pylint --exit-zero xcp/ tests/ setup.py
           pylint --exit-zero --msg-template="{path}:{line}: [{msg_id}({symbol}), {obj}] {msg}" xcp/ tests/ setup.py > pylint.txt
-          diff-quality --violations=pylint --html-report pylint-diff.html pylint.txt
+          diff-quality --compare-branch=origin/master --violations=pylint --html-report pylint-diff-${{ matrix.python-version }}.html pylint.txt
+
+      - name: Get Cover
+        if: github.event_name == 'pull_request' && steps.coverage.outcome == 'success'
+        uses: orgoro/coverage@v3
+        with:
+          # https://github.com/marketplace/actions/python-coverage
+          coverageFile: coverage.xml
+          thresholdNew: 0.80
+          thresholdAll: 0.67
+          thresholdModified: 0.20
+          token: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Upload coverage reports to Codecov
+        uses: codecov/codecov-action@v3
+        if: github.event_name == 'pull_request' && steps.coverage.outcome == 'success'
```

Comment:
Don't we want code coverage for all commits, not just pull requests?

Reply:
That means the action has to be re-enabled run for all pushes, fine if you want!

I'll re-enable it and see how it goes.